### PR TITLE
Port twisted.protocols.finger to Python 3

### PR DESCRIPTION
**Pull request by rodrigc**
http://twistedmatrix.com/trac/ticket/8230

### CODE REVIEW

**Comment by rodrigc** [comment]
@leth thanks for your previous work on Twisted.  Can you code review this on GitHub?

**Comment by leth** [comment]
The diff looks reasonable :) Just a couple of minor subjective comments!

I note there's no decoding going on; from an API design point of view, does it make sense for this API to be returning bytes?

File: docs/core/howto/tutorial/listings/finger/fingerXRclient.py [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
 
-from __future__ import print_function
+try:
+    # Python 3
+    import xmlrpc.client
+    Server = xmlrpc.client.Server
+except ImportError:
+    # Python 2
+    import xmlrpclib
+    Server = xmlrpclib.Server
```
Comment (leth) [inline docs/core/howto/tutorial/listings/finger/fingerXRclient.py]: You could use "from" imports here

**Comment by leth** [comment]
There are a couple of places with docstrings which say "strings" e.g. `IFingerSetterFactory`, which will be at odds with python3 if things are bytes.

File: docs/core/howto/tutorial/listings/finger/fingerXRclient.py (hunk 1 above)
Comment (rodrigc) [reply docs/core/howto/tutorial/listings/finger/fingerXRclient.py]: OK

**Comment by rodrigc** [comment]
@leth the finger protocol itself is quite old, and deals with bytes.  The code in Twisted is also quite old, since it was one of the first protocols written.  Now, not a lot of people are going to be implementing finger servers these days.  However, there is a nice Twisted tutorial based on this protocol, and it would be really nice if the examples worked more or less out of the box on Python 3,
to give new users a nice experience, instead of running into a lot of errors.

Based on your comments, I took a pass through the tutorial and changed the docstring for things like `IFingerSetterFactory` to bytes.  I also fixed up some of the examples in the finger tutorial.  Before, none of the examples worked on Python 3, but now quite a lot of them do.

I didn't get to all of them, but if it is OK with you, I'd like to commit this stuff to begin with,
and either I or other people can fix up the other examples over time.

Thanks!

**Comment by leth** [comment]
Sounds reasonable!

On 29 January 2017 22:48:07 GMT+00:00, Craig Rodrigues <notifications@github.com> wrote:
>@leth the finger protocol itself is quite old, and deals with bytes. 
>The code in Twisted is also quite old, since it was one of the first
>protocols written.  Now, not a lot of people are going to be
>implementing finger servers these days.  However, there is a nice
>Twisted tutorial based on this protocol, and it would be really nice if
>the examples worked more or less out of the box on Python 3,
>to give new users a nice experience, instead of running into a lot of
>errors.
>
>Based on your comments, I took a pass through the tutorial and changed
>the docstring for things like `IFingerSetterFactory` to bytes.  I also
>fixed up some of the examples in the finger tutorial.  Before, none of
>the examples worked on Python 3, but now quite a lot of them do.
>
>I didn't get to all of them, but if it is OK with you, I'd like to
>commit this stuff to begin with,
>and either I or other people can fix up the other examples over time.
>
>Thanks!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/twisted/twisted/pull/689#issuecomment-275952841

-- 
Sent from my phone. Please excuse my brevity.

**Comment by rodrigc** [comment]
@leth can you mark this review as approved on GitHub so I can list you as a reviewer in the commit message?

**Comment by rodrigc** [comment]
Also, just out of curiousity, do you work on @einarfd 's team on this: https://github.com/twisted/twisted/pull/456#issuecomment-242096969

**Review by leth — APPROVED** [review]

**Comment by leth** [comment]
Not quite, we're in the same arm of the megacorp but different parts of it.
Our two groups have a history of collaboration; we have an internal project built on twisted which has an open-source-like collaboration model.

**Comment by rodrigc** [comment]
@moshez thanks for writing this tutorial a long time ago.  I updates some of the examples to work in Python 3. :)

**Pull request merged**